# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: dom/workers/remoteworkers/RemoteWorkerService.cpp
# Commit: 1eeb386120ff
# Full Hash: 1eeb386120ff5b10f793f9ad7cdff1675a5c1ef1
# Author: Eden Chuang <echuang@mozilla.com>
# Date: 2025-04-29 20:49:47
# Description:
#   Bug 1961460 - Remove the assertion for target thread in RemoteWorkerDebuggerManagerChild::Constructor. r=dom-worker-reviewers,jstutte
#   
#   RemoteWorkerService::InitializeOnTargetThread could race with its shutdown
#   So we can not use sRemoteWorkerService for thread correctness checking in RemoteWorkerDebuggerManagerChild creation during the initialization.
#   
# ==============================================================================

diff -r 9138373fb84a -r 1eeb386120ff dom/workers/remoteworkers/RemoteWorkerService.cpp
--- a/dom/workers/remoteworkers/RemoteWorkerService.cpp	Tue Apr 29 13:26:40 2025 +0000
+++ b/dom/workers/remoteworkers/RemoteWorkerService.cpp	Tue Apr 29 13:50:20 2025 +0000
@@ -275,11 +275,8 @@
       "InitializeThread",
       [self, endpoint = std::move(aEndpoint),
        debuggerChildEp = std::move(aDebuggerChildEp)]() mutable {
-        self->InitializeOnTargetThread(std::move(endpoint));
-
-        self->mDebuggerManagerChild =
-            MakeRefPtr<RemoteWorkerDebuggerManagerChild>();
-        debuggerChildEp.Bind(self->mDebuggerManagerChild);
+        self->InitializeOnTargetThread(std::move(endpoint),
+                                       std::move(debuggerChildEp));
       });
 
   rv = mThread->Dispatch(r.forget(), NS_DISPATCH_NORMAL);
@@ -298,18 +295,26 @@
 RemoteWorkerService::~RemoteWorkerService() = default;
 
 void RemoteWorkerService::InitializeOnTargetThread(
-    mozilla::ipc::Endpoint<PRemoteWorkerServiceChild> aEndpoint) {
+    mozilla::ipc::Endpoint<PRemoteWorkerServiceChild> aEndpoint,
+    mozilla::ipc::Endpoint<PRemoteWorkerDebuggerManagerChild>
+        aDebuggerMgrEndpoint) {
   MOZ_ASSERT(mThread);
   MOZ_ASSERT(mThread->IsOnCurrentThread());
 
+  RefPtr<RemoteWorkerDebuggerManagerChild> debuggerManagerActor =
+      MakeRefPtr<RemoteWorkerDebuggerManagerChild>();
+  if (NS_WARN_IF(!aDebuggerMgrEndpoint.Bind(debuggerManagerActor))) {
+    return;
+  }
+
   RefPtr<RemoteWorkerServiceChild> serviceActor =
       MakeAndAddRef<RemoteWorkerServiceChild>();
   if (NS_WARN_IF(!aEndpoint.Bind(serviceActor))) {
     return;
   }
 
-  // Now we are ready!
-  mActor = serviceActor;
+  mDebuggerManagerChild = std::move(debuggerManagerActor);
+  mActor = std::move(serviceActor);
 }
 
 void RemoteWorkerService::CloseActorOnTargetThread() {
@@ -322,6 +327,10 @@
     mActor->Close();
     mActor = nullptr;
   }
+  if (mDebuggerManagerChild) {
+    mDebuggerManagerChild->Close();
+    mDebuggerManagerChild = nullptr;
+  }
 }
 
 NS_IMETHODIMP